service/storage_service.cc

```diff
@@ -3551,7 +3551,7 @@ class node_ops_ctl {
             seastar::thread::maybe_yield();
             // FIXME: use node* rather than endpoint
             auto node = np->endpoint();
-            if (!ignore_nodes.contains(node) && sync_to_node(node)) {
+            if (!ignore_nodes.contains(node) && tmptr->is_normal_token_owner(node) && sync_to_node(node)) {
```

Comment:
What nodes are stored in get_topology()?   What's the point of using get_topology::for_each_node() and filter with tmptr->is_normal_token_owner? We already have tmptr->get_all_endpoints() to return all normal token owners nodes?




```
    // Returns nodes that are officially part of the ring. It does not include
    // node that is still joining the cluster, e.g., a node that is still
    // streaming data before it finishes the bootstrap process and turns into
    // NORMAL status.
    const std::unordered_set<inet_address>& get_all_endpoints() const noexcept {
        return _normal_token_owners;
    }
```

Reply:
The topology may hold pending and leaving nodes as well as left nodes.
The reason to use topology rather than `get_all_endpoints()` is the transition to use topology::node that has both the ip address and host_id (as well as dc/rack).
The end game indeed is planned be to use nodes/host_ids in token_metadata so `get_all_endpoints` or a similar function would be able to return a set of nodes instead of endpoints, but we're not there yet.
But it might be a good idea to provide this interface now, e.g. `get_all_nodes()` that can construct a unordered_set of nodes which are normal token owners.

Reply:
Are you talking about making tm::get_all_nodes return a set of locator::node?

BTW, we even have locator::node::_state. It makes less sense to query it again with tmptr->is_normal_token_owner(node) instead of checking _state directly. Is _state accurate enough about the state? Why not check the flag directly?

```
enum class state {                                             
    none = 0,                                                  
    joining,    // while bootstrapping, replacing              
    normal,                                                    
    leaving,    // while decommissioned, removed, replaced     
    left        // after decommissioned, removed, replaced     
};                  
```

Reply:
The node state should be correct, but it `leaving` doesn't distinguish between decommissioned nodes to removed nodes.   Maybe it's a good opportunity to make this distinction if possible and here we'd need to return all nodes in `normal` or `decommissioning` state.

Reply:
> Are you talking about making tm::get_all_nodes return a set of locator::node?

Yes.